compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java

```diff
@@ -5855,10 +5756,80 @@ public void visit(BLangGroupExpr groupExpr) {
     @Override
     public void visit(BLangRecordLiteral recordLiteral) {
         List<RecordLiteralNode.RecordField> fields =  recordLiteral.fields;
+        generateFieldsForUserUnspecifiedRecordFields(recordLiteral, fields);
         fields.sort((v1, v2) -> Boolean.compare(isComputedKey(v1), isComputedKey(v2)));
         result = rewriteExpr(rewriteMappingConstructor(recordLiteral));
     }
 
+    @Override
+    public void visit(BFunctionPointerInvocation functionPointerInvocation) {
+        result = functionPointerInvocation;
+    }
+
+    private List<String> getNamesOfUserSpecifiedRecordFields(List<RecordLiteralNode.RecordField> userSpecifiedFields) {
+        List<String> fieldNames = new ArrayList<>();
+        for (RecordLiteralNode.RecordField field : userSpecifiedFields) {
+            if (field.getKind() == NodeKind.RECORD_LITERAL_KEY_VALUE) {
+                BLangExpression key = ((BLangRecordLiteral.BLangRecordKeyValueField) field).key.expr;
+                if (key.getKind() == NodeKind.LITERAL) {
+                    fieldNames.add(((BLangLiteral) key).originalValue);
+                } else if (key.getKind() == NodeKind.SIMPLE_VARIABLE_REF) {
+                    fieldNames.add(((BLangSimpleVarRef) key).variableName.value);
+                }
+            }
+        }
+        return fieldNames;
+    }
+
+    private void generateFieldsForUserUnspecifiedRecordFields(List<RecordLiteralNode.RecordField> fields,
+                                                              List<String> fieldNames,
+                                                              Map<String, BInvokableSymbol> defaultValues,
+                                                              Location pos, boolean isReadonly) {
+        for (Map.Entry<String, BInvokableSymbol> entry : defaultValues.entrySet()) {
+            String fieldName = entry.getKey();
+            if (fieldNames.contains(fieldName)) {
+                continue;
+            }
+            fieldNames.add(fieldName);
+            BInvokableSymbol invokableSymbol = entry.getValue();
+            BLangExpression expression = getInvocation(invokableSymbol);
+            if (isReadonly && !Symbols.isFlagOn(invokableSymbol.retType.flags, Flags.READONLY)) {
+                expression = visitCloneReadonly(expression, invokableSymbol.retType);
```

Comment:
Consider

```ballerina
type Foo record {|
    any[] x = [1, new stream<int>()];
|};

public function main() {
    Foo & readonly _ = {};
}
```

This results in a panic at runtime atm.

```cmd
error: java.lang.UnsupportedOperationException
        at ballerina.lang.value.0:cloneReadOnly(value.bal:58)
           y:main(y.bal:6)
```

This isn't correct?

Reply:
In https://github.com/ballerina-platform/ballerina-spec/issues/544#issuecomment-1064777468, James has said

> Then I guess we say that in the readonly case if the type of the default expression is not Cloneable then there is no default. 

Please follow up on the issue if required.

Btw, if we are to consider this scenario as one with no default, we need to have this default value information at semantic analysis.

Reply:
What is the plan for this?

Reply:
Looking into this

Reply:
https://github.com/ballerina-platform/ballerina-spec/issues/544#issuecomment-1598057258

Reply:
Fixed with https://github.com/ballerina-platform/ballerina-lang/pull/37184/commits/bddee51f61fd88f2147c7c29a072ad5bcd98bc3c